doc/source/reference/arrays.classes.rst

```diff
@@ -39,77 +39,88 @@ Special attributes and methods
-
-   .. note:: If you also define right-hand binary operator override
-      methods (such as ``__rmul__``) or comparison operations (such as
-      ``__gt__``) in your class, they take precedence over the
-      :func:`__numpy_ufunc__` mechanism when resolving results of
-      binary operations (such as ``ndarray_obj * your_obj``).
-
-      The technical special case is: ``ndarray.__mul__`` returns
-      ``NotImplemented`` if the other object is *not* a subclass of
-      :class:`ndarray`, and defines both ``__numpy_ufunc__`` and
-      ``__rmul__``. Similar exception applies for the other operations
-      than multiplication.
-
-      In such a case, when computing a binary operation such as
-      ``ndarray_obj * your_obj``, your ``__numpy_ufunc__`` method
-      *will not* be called.  Instead, the execution passes on to your
-      right-hand ``__rmul__`` operation, as per standard Python
-      operator override rules.
-
-      Similar special case applies to *in-place operations*: If you
-      define ``__rmul__``, then ``ndarray_obj *= your_obj`` *will not*
-      call your ``__numpy_ufunc__`` implementation. Instead, the
-      default Python behavior ``ndarray_obj = ndarray_obj * your_obj``
-      occurs.
-
-      Note that the above discussion applies only to Python's builtin
-      binary operation mechanism. ``np.multiply(ndarray_obj,
-      your_obj)`` always calls only your ``__numpy_ufunc__``, as
-      expected.
-
-.. method:: class.__array_finalize__(obj)
+   .. note:: In addition to ufuncs, :func:`__array_ufunc__` also
+      overrides the behavior of :func:`numpy.dot` and :func:`numpy.matmul`.
+      This even though these are not ufuncs, but they can be thought of as
+      :ref:`generalized universal functions<c-api.generalized-ufuncs>`
+      (which are overridden). We intend to extend this behaviour to other
+      relevant functions.
+
+   Like with other methods in python, such as ``__hash__`` and
+   ``__iter__``, it is possible to indicate that your class does *not*
+   support ufuncs by setting ``__array_ufunc__ = None``. With this,
+   inside ufuncs, your class will be treated as if it returned
+   :obj:`NotImplemented` (which will lead to an :exc:`TypeError`
+   unless another class also provides a :func:`__array_ufunc__` method
+   which knows what to do with your class).
+
+   The presence of :func:`__array_ufunc__` also influences how binary
+   and comparison operators are dealt with, such as ``__add__``,
+   ``__gt__``, etc.  If it is not :obj:`None`, the assumption is that
+   your code can handle such operations via the ufunc mechanism, and
+   hence forward methods on :class:`ndarray` will call the ufuncs
+   unconditionally (i.e., even if your class has defined reverse
+   methods such as ``__radd__``, ``__le__``, etc.). If
+   ``__array_ufunc__ = None``, however, forward methods on
+   :class:`ndarray` will unconditionally return :obj:`NotImplemented`,
+   so that your reverse methods will get called.
```

Comment:
The use of the undefined term "forward methods" here is pretty confusing – maybe it should be something like:

The presence of `__array_ufunc__` also influences how `ndarray` handles special binary operations like `arr + obj` and `arr < obj` when `arr` is an `ndarray` and `obj` is an instance of a custom class. There are two possibilities. If `obj.__array_ufunc__` is present and not `None`, then `ndarray.__add__` and friends will delegate to the ufunc machinery, meaning that `arr + obj` becomes `np.add(arr, obj)`, and then `np.add` invokes `__array_ufunc__`. This is useful if you want to define an object that acts like an array. If your class defines `__array_ufunc__`, then for consistency it should also define special methods like `__add__` and `__lt__` that delegate to ufuncs just like ndarray does. [Do we have a helper for this?]

Alternatively, if `obj.__array_ufunc__` is set to `None`, then as a special case, special methods like `ndarray.__add__` will notice this and *unconditionally* return `NotImplemented`, so that Python will dispatch to `obj.__radd__` instead. This is useful if you want to define a special object that interacts with arrays via binary operations, but isn't itself an array. For example, a units handling system might have an object `m` representing the "meters" unit, and want to support the syntax `arr * m` to represent that the array has units of "meters", but not want to otherwise interact with arrays via ufuncs or otherwise. This can be done by setting `__array_ufunc__ = None` and defining `__mul__` and `__rmul__` methods. (Note that this means that writing an `__array_ufunc__` that always returns `NotImplemented` is not quite the same as setting `__array_ufunc__ = None`: in the former case, `arr + obj` will raise `TypeError`, while in the latter case it's possible to define a `__radd__` method to prevent this.)